routetable/route_table.go

```diff
@@ -426,88 +533,203 @@ func (r *RouteTable) ensureL2Dataplane(linkAttrs *netlink.LinkAttrs, target L2Ta
 	}
 
-	// The code below may add some more CIDRs to clean up before it is done, make sure we
-	// remove conntrack entries in any case.
-	defer oldCIDRs.Iter(func(item interface{}) error {
-		// Remove and conntrack entries that should no longer be there.
-		dest := item.(ip.CIDR)
-		r.startConntrackDeletion(dest.Addr())
-		return nil
-	})
+	// Any entry that we delete, we should remove the corresponding conntrack entries.
+	defer func() {
+		for cidr := range deletedCIDRsToTarget {
+			r.startConntrackDeletion(cidr.Addr())
+		}
+	}()
 
 	// Try to get the link.  This may fail if it's been deleted out from under us.
-	nl, err := r.getNetlinkHandle()
+	linkAttrs, err := r.getLinkAttributes(ifaceName)
 	if err != nil {
-		r.logCxt.WithError(err).Error("Failed to connect to netlink, retrying...")
-		return ConnectFailed
+		return err
 	}
-	link, err := nl.LinkByName(ifaceName)
-	if err != nil {
-		// Filter the error so that we don't spam errors if the interface is being torn
-		// down.
-		filteredErr := r.filterErrorByIfaceState(ifaceName, err, GetFailed)
-		if filteredErr == GetFailed {
-			logCxt.WithError(err).Error("Failed to get interface.")
-			r.closeNetlinkHandle() // Defensive: force a netlink reconnection next time.
-		} else {
-			logCxt.WithError(err).Info("Failed to get interface; it's down/gone.")
+
+	if !fullSync {
+		// If we aren't doing a full re-sync then attempt to apply deltas to the route table.
+		logCxt.Debug("Perform a delta route sync")
+		if err := r.deltaUpdateRoutesForLink(logCxt, linkAttrs, ifaceName, deletedCIDRsToTarget, createdCIDRsTarget); err != UpdateFailed {
+			return err
 		}
-		return filteredErr
 	}
 
+	// We are either scheduled to do a full resync, or a previous update failed. In either case, attempt a full
+	// resync of state.
+	logCxt.Debugf("Perform a full route sync: %v %v", updatedCIDRsToTarget, deletedCIDRsToTarget)
+	return r.fullSyncRoutesForLink(logCxt, linkAttrs, ifaceName, updatedCIDRsToTarget, deletedCIDRsToTarget)
+}
+
+func (r *RouteTable) createL3Route(linkAttrs *netlink.LinkAttrs, target Target) netlink.Route {
+	var linkIndex int
+	if linkAttrs != nil {
+		linkIndex = linkAttrs.Index
+	}
+	cidr := target.CIDR
+	ipNet := cidr.ToIPNet()
+	route := netlink.Route{
+		LinkIndex: linkIndex,
+		Dst:       &ipNet,
+		Type:      target.RouteType(),
+		Protocol:  r.deviceRouteProtocol,
+		Scope:     target.RouteScope(),
+		Table:     r.tableIndex,
+	}
+
+	if r.deviceRouteSourceAddress != nil {
+		route.Src = r.deviceRouteSourceAddress
+	}
+
+	if target.GW != nil {
+		route.Gw = target.GW.AsNetIP()
+	}
+
+	if target.Type == TargetTypeVXLAN || target.Type == TargetTypeNoEncap {
+		route.Scope = netlink.SCOPE_UNIVERSE
+		route.SetFlag(syscall.RTNH_F_ONLINK)
+	}
+
+	return route
+}
+
+// fullSyncRoutesForLink performs a full resync of the routes by first listing current routes and correlating against
+// the expected set.
+func (r *RouteTable) fullSyncRoutesForLink(
```

Comment:
Could the "full resync" case and the "delta updates" case be combined to avoid having two fairly complex codepaths to maintain?

Elsewhere, I've used this pattern to force both cases through the same logic:

* Track our complete, "desired state" of the dataplane
* Track which parts of the dataplane we think are dirty
* When making a delta change, update desired state and mark everything touched as dirty
* When doing a resync, throw away the dirty set and recalculate it from scratch; the resync doesn't actually modify the dataplane, only our picture of what's dirty
* Then (whether or not we did a resync) to apply updates, iterate over the dirty set and queue up an update to bring the dataplane into sync with the desired state.  If the resync found everything was already in sync, the dirty set ends up empty and we do nothing at this step.

Example: https://github.com/projectcalico/felix/blob/master/dataplane/linux/bpf_route_mgr.go#L124